core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -73,6 +105,40 @@ public void request(int numMessages) {
   public void sendHeaders(Metadata headers) {
     checkState(!sendHeadersCalled, "sendHeaders has already been called");
     checkState(!closeCalled, "call is closed");
+
+    headers.removeAll(MESSAGE_ENCODING_KEY);
+    if (compressor == null) {
+      compressor = Codec.Identity.NONE;
+      if (inboundHeaders.containsKey(MESSAGE_ACCEPT_ENCODING_KEY)) {
+        String acceptEncodings = inboundHeaders.get(MESSAGE_ACCEPT_ENCODING_KEY);
+        for (String acceptEncoding : ACCEPT_ENCODING_SPLITER.split(acceptEncodings)) {
+          Compressor c = compressorRegistry.lookupCompressor(acceptEncoding);
+          if (c != null) {
+            compressor = c;
+            break;
+          }
+        }
+      }
+    } else {
+      if (inboundHeaders.containsKey(MESSAGE_ACCEPT_ENCODING_KEY)) {
+        String acceptEncodings = inboundHeaders.get(MESSAGE_ACCEPT_ENCODING_KEY);
+        List<String> acceptedEncodingsList = ACCEPT_ENCODING_SPLITER.splitToList(acceptEncodings);
+        checkArgument(acceptedEncodingsList.contains(compressor.getMessageEncoding()),
```

Comment:
Doesn't this go against the spec again? 
> For every message a server is requested to compress using an algorithm it knows the client doesn't support (as indicated by the last grpc-accept-encoding header received from the client), it SHALL send the message uncompressed.

Also, if the client doesn't send any accept encoding then we should also not use compression.